# notebook info

In this notebook parsed dataframe will be adjusted to the kaggle test dataframe format. The following changes have been made:
- replaced feature names
- replace samo values, because description in train dataframe defer from the ets dataframe
- created some new features based on the information in test data frmae like `fuel type, 


In [81]:
import numpy as np  # linear algebra
import pandas as pd
from progressbar import ProgressBar
pbar = ProgressBar()

In [82]:
# read  multiple dataframes
# create a list with file names
files = os.listdir('data')
files.remove('.DS_Store')

# create empty dataframe to concat
column_names = ['brand', 'model_name', 'price', 'год выпуска', 'Пробег', 'Кузов', 'Цвет', 'Двигатель', 'Налог',
                'Коробка', 'Привод', 'Руль', 'Состояние', 'Владельцы', 'ПТС', 'Таможня', 'VIN', 'Госномер', 'comment']

train = pd.DataFrame(columns=column_names)

# loop to concat datafrmaes
for file in files:
    temp = pd.read_csv(os.path.join(
        '/Users/vramich/Documents/SF/Unit_7_ML_Basics_partII/project_06/data', file))
    train = pd.concat([train, temp])


train = train.drop_duplicates()
print('train length', len(train))
# reading test
test = pd.read_csv('test.csv')
print('test length: ', len(test))

train length 89712
test length:  34686


In [83]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 89712 entries, 0 to 846
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   brand        89712 non-null  object 
 1   model_name   89671 non-null  object 
 2   price        89670 non-null  object 
 3   год выпуска  89669 non-null  object 
 4   Пробег       89669 non-null  object 
 5   Кузов        89666 non-null  object 
 6   Цвет         89668 non-null  object 
 7   Двигатель    89666 non-null  object 
 8   Налог        88700 non-null  object 
 9   Коробка      89666 non-null  object 
 10  Привод       89666 non-null  object 
 11  Руль         89668 non-null  object 
 12  Состояние    89668 non-null  object 
 13  Владельцы    89663 non-null  object 
 14  ПТС          89666 non-null  object 
 15  Таможня      89668 non-null  object 
 16  VIN          80885 non-null  object 
 17  Госномер     61870 non-null  object 
 18  comment      89660 non-null  object 
 19  Владен

In [84]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34686 entries, 0 to 34685
Data columns (total 32 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   bodyType              34686 non-null  object
 1   brand                 34686 non-null  object
 2   car_url               34686 non-null  object
 3   color                 34686 non-null  object
 4   complectation_dict    6418 non-null   object
 5   description           34686 non-null  object
 6   engineDisplacement    34686 non-null  object
 7   enginePower           34686 non-null  object
 8   equipment_dict        24690 non-null  object
 9   fuelType              34686 non-null  object
 10  image                 34686 non-null  object
 11  mileage               34686 non-null  int64 
 12  modelDate             34686 non-null  int64 
 13  model_info            34686 non-null  object
 14  model_name            34686 non-null  object
 15  name                  34686 non-null

In [85]:
train.head(2)

brand                    model_name     price год выпуска      Пробег  \
0   bmw   BMW 5 серии  III (E34) 525i   155000₽        1988  226 000 км   
1   bmw  BMW X6 M  I (E71) Рестайлинг  1950000₽        2013   75 000 км   

               Кузов    Цвет                  Двигатель           Налог  \
0              седан  чёрный  2.5 л / 170 л.с. / Бензин   8 500 ₽ / год   
1  внедорожник 5 дв.   белый  4.4 л / 555 л.с. / Бензин  83 250 ₽ / год   

          Коробка  ...                VIN    Госномер  \
0    механическая  ...  WBA**************  ******|174   
1  автоматическая  ...  WBS**************   ******|74   

                                             comment           Владение  \
0  Продаю авто, за время владения было не мало чт...                NaN   
1  Дилерский авто , в матовой  защитной пленке  (...  6 лет и 9 месяцев   

  Гарантия Обмен Кузов № Запас хода Статус Unnamed: 0  
0      NaN   NaN     NaN        NaN    NaN        NaN  
1      NaN   NaN     NaN        NaN    NaN        NaN  

[2 rows x 26 columns]

In [86]:
test.head(2)

bodyType  brand                                            car_url   color  \
0  лифтбек  SKODA  https://auto.ru/cars/used/sale/skoda/octavia/1...   синий   
1  лифтбек  SKODA  https://auto.ru/cars/used/sale/skoda/octavia/1...  чёрный   

  complectation_dict                                        description  \
0                NaN  Все автомобили, представленные в продаже, прох...   
1                NaN  ЛОТ: 01217195\nАвтопрага Север\nДанный автомоб...   

  engineDisplacement enginePower  \
0            1.2 LTR     105 N12   
1            1.6 LTR     110 N12   

                                      equipment_dict fuelType  ...  \
0  {"engine-proof":true,"tinted-glass":true,"airb...   бензин  ...   
1  {"cruise-control":true,"asr":true,"esp":true,"...   бензин  ...   

      vehicleConfiguration  vehicleTransmission    vendor    Владельцы  \
0       LIFTBACK ROBOT 1.2     роботизированная  EUROPEAN  3 или более   
1  LIFTBACK MECHANICAL 1.6         механическая  EUROPEAN   1 владелец   

  Владение       ПТС    Привод   Руль           Состояние     Таможня  
0      NaN  Оригинал  передний  Левый  Не требует ремонта  Растаможен  
1      NaN  Оригинал  передний  Левый  Не требует ремонта  Растаможен  

[2 rows x 32 columns]

In [89]:
# a) rename columns
# print columns names in train
train.columns

Index(['brand', 'model_name', 'price', 'год выпуска', 'Пробег', 'Кузов',
       'Цвет', 'Двигатель', 'Налог', 'Коробка', 'Привод', 'Руль', 'Состояние',
       'Владельцы', 'ПТС', 'Таможня', 'VIN', 'Госномер', 'comment', 'Владение',
       'Гарантия', 'Обмен', 'Кузов №', 'Запас хода', 'Статус', 'Unnamed: 0'],
      dtype='object')

In [ ]:
['brand', 'model_name', 'price', 'productionDate', 'mileage', 'bodyType',
 'color', 'engineDisplacement', 'Налог', 'vehicleTransmission', 'Привод',
 'Руль', 'Состояние', 'Владельцы', 'ПТС', 'Таможня', 'VIN', 'Госномер',
 'description', 'Владение', 'Обмен', 'Гарантия', 'Кузов №', 'Запас хода',
 'Статус', 'enginePower', 'fuelType', 'vendor']

## engineDisplacement 
b) has following values: 1.6 л / 110 л.с. / Бензин
where:
- `1.6 л` -> **engineDisplacement**
- `110 л.с.`-> **enginePower**
- `Бензин` -> **fuelType**

c) We will split engineDisplacement into three columns descrivbed above and bring the values into the same format:

- relplace `л` through `LTR` 
- replace `л.с.` through `N12`

In [90]:
train.rename(columns={'год выпуска': 'productionDate', 'Пробег': 'mileage', 'Кузов': 'bodyType', 'Цвет': 'color',
                      'Двигатель': 'engineDisplacement', 'Коробка': 'vehicleTransmission', 'comment': 'description'},
             inplace=True)
train.head()

brand                              model_name     price productionDate  \
0   bmw             BMW 5 серии  III (E34) 525i   155000₽           1988   
1   bmw            BMW X6 M  I (E71) Рестайлинг  1950000₽           2013   
2   bmw  BMW X1  II (F48) Рестайлинг 18d xDrive  2800000₽           2020   
3   bmw                     BMW X4  I (F26) 28i  2950000₽           2018   
4   bmw                    BMW X5  IV (G05) 30d  5450000₽           2019   

      mileage           bodyType   color         engineDisplacement  \
0  226 000 км              седан  чёрный  2.5 л / 170 л.с. / Бензин   
1   75 000 км  внедорожник 5 дв.   белый  4.4 л / 555 л.с. / Бензин   
2   18 565 км  внедорожник 5 дв.   белый  2.0 л / 150 л.с. / Дизель   
3   21 000 км  внедорожник 5 дв.   серый  2.0 л / 245 л.с. / Бензин   
4   78 200 км  внедорожник 5 дв.  чёрный  3.0 л / 249 л.с. / Дизель   

            Налог vehicleTransmission  ...                VIN    Госномер  \
0   8 500 ₽ / год        механическая  ...  WBA**************  ******|174   
1  83 250 ₽ / год      автоматическая  ...  WBS**************   ******|74   
2   2 010 ₽ / год      автоматическая  ...  X4X**************  ******|174   
3  18 375 ₽ / год      автоматическая  ...  X4X**************  ******|174   
4  20 543 ₽ / год      автоматическая  ...  WBA**************  ******|174   

                                         description           Владение  \
0  Продаю авто, за время владения было не мало чт...                NaN   
1  Дилерский авто , в матовой  защитной пленке  (...  6 лет и 9 месяцев   
2  Активная безопасность:— Антиблокировочная сист...                NaN   
3  БМВ-Х4. Бензин -245л.сил. АКПП-8. Куплен у ОФ....                NaN   
4  Куплен и oбcлуживaлся у официального дилерa. О...   1 год и 2 месяца   

          Гарантия Обмен Кузов № Запас хода Статус Unnamed: 0  
0              NaN   NaN     NaN        NaN    NaN        NaN  
1              NaN   NaN     NaN        NaN    NaN        NaN  
2              NaN   NaN     NaN        NaN    NaN        NaN  
3              NaN   NaN     NaN        NaN    NaN        NaN  
4  До декабря 2022   NaN     NaN        NaN    NaN        NaN  

[5 rows x 26 columns]

In [91]:
# b)
train[['engineDisplacement', 'enginePower', 'fuelType']] = (  # Create two new features
    train["engineDisplacement"]           # from the engineDisplacement feature
    .str                         # through the string accessor
    .split("/", expand=True)     # by splitting on "/"
                                 # and expanding the result into separate columns
)
# check new columns
train.head(2)

brand                    model_name     price productionDate     mileage  \
0   bmw   BMW 5 серии  III (E34) 525i   155000₽           1988  226 000 км   
1   bmw  BMW X6 M  I (E71) Рестайлинг  1950000₽           2013   75 000 км   

            bodyType   color engineDisplacement           Налог  \
0              седан  чёрный             2.5 л    8 500 ₽ / год   
1  внедорожник 5 дв.   белый             4.4 л   83 250 ₽ / год   

  vehicleTransmission  ...                                        description  \
0        механическая  ...  Продаю авто, за время владения было не мало чт...   
1      автоматическая  ...  Дилерский авто , в матовой  защитной пленке  (...   

            Владение Гарантия Обмен Кузов № Запас хода Статус Unnamed: 0  \
0                NaN      NaN   NaN     NaN        NaN    NaN        NaN   
1  6 лет и 9 месяцев      NaN   NaN     NaN        NaN    NaN        NaN   

  enginePower fuelType  
0   170 л.с.    Бензин  
1   555 л.с.    Бензин  

[2 rows x 28 columns]

In [92]:
# c)

train['engineDisplacement'] = train['engineDisplacement'].str.replace(' ', '')
train['engineDisplacement'] = train['engineDisplacement'].str.replace(
    'л', ' LTR')

train['enginePower'] = train['enginePower'].str.replace(' ', '')
train['enginePower'] = train['enginePower'].str.replace('л.с.', ' N12')

# compare features
`TRAIN`:
- we have 9 unmatched features
- all the features have a very big number of empty values and doesn't brin much insoghts into the car characteristics
> delete this 9 featurs as it is better to have no data rather that have wrong data

`TEST`:
- in test dataset we have 13 unmatched features
- name & enginePower have the same information
- model_name contain model_info information
- `vehicleConfiguration` contains important car characteristics. Unfortunately I didn't parsed that information for train dataset and have to delete this feature 

In [93]:
# check the feature we have in test and don't have in train dataset
train.columns.difference(test.columns)

Index(['Unnamed: 0', 'VIN', 'price', 'Гарантия', 'Госномер', 'Запас хода',
       'Кузов №', 'Налог', 'Обмен', 'Статус'],
      dtype='object')

In [94]:
# check this variables
train[['VIN', 'price', 'Гарантия', 'Госномер', 'Запас хода',
       'Кузов №', 'Налог', 'Обмен', 'Статус']].describe(include='all')

VIN    price       Гарантия    Госномер Запас хода  \
count               80885    89670           7496       61870        298   
unique                986     6199             93         236         18   
top     XW8**************  450000₽  До марта 2023  ******|799     199 км   
freq                10724      884           1742        3394        122   

                Кузов №           Налог               Обмен    Статус  
count              7904           88700               31011        27  
unique             2106            2445                   1         1  
top     FB15***********  18 675 ₽ / год  Рассмотрю варианты  На заказ  
freq                138            3370               31011        27

In [95]:
# check number of NA's
train[['VIN', 'price', 'Гарантия', 'Госномер', 'Запас хода', 'Кузов №',
       'Налог', 'Обмен', 'Статус', 'Unnamed: 0']].isnull().sum()

VIN            8827
price            42
Гарантия      82216
Госномер      27842
Запас хода    89414
Кузов №       81808
Налог          1012
Обмен         58701
Статус        89685
Unnamed: 0    81925
dtype: int64

In [96]:
# check the feature we have in test and don't have in train dataset
test.columns.difference(train.columns)

Index(['car_url', 'complectation_dict', 'equipment_dict', 'image', 'modelDate',
       'model_info', 'name', 'numberOfDoors', 'parsing_unixtime',
       'priceCurrency', 'sell_id', 'super_gen', 'vehicleConfiguration',
       'vendor'],
      dtype='object')

In [97]:
# check this variables
test[['car_url', 'complectation_dict', 'equipment_dict', 'image', 'modelDate',
      'model_info', 'name', 'numberOfDoors', 'parsing_unixtime',
      'priceCurrency', 'sell_id', 'super_gen', 'vehicleConfiguration']].describe(include='all').T

count unique  \
car_url               34686  34686   
complectation_dict     6418   2364   
equipment_dict        24690  23705   
image                 34686  34557   
modelDate             34686    NaN   
model_info            34686    954   
name                  34686   2780   
numberOfDoors         34686    NaN   
parsing_unixtime      34686    NaN   
priceCurrency         34686      1   
sell_id               34686    NaN   
super_gen             34686   5890   
vehicleConfiguration  34686    634   

                                                                    top  \
car_url               https://auto.ru/cars/used/sale/volvo/xc90/1100...   
complectation_dict    {"id":"4562904","name":"Elegance","available_o...   
equipment_dict                                         {"leather":true}   
image                 https://avatars.mds.yandex.net/get-verba/21620...   
modelDate                                                           NaN   
model_info            {"code":"OCTAVIA","name":"Octavia","ru_name":"...   
name                                                  1.6 AT (110 л.с.)   
numberOfDoors                                                       NaN   
parsing_unixtime                                                    NaN   
priceCurrency                                                       RUB   
sell_id                                                             NaN   
super_gen             {"id":"6214876","displacement":1598,"engine_ty...   
vehicleConfiguration                      ALLROAD_5_DOORS AUTOMATIC 3.0   

                       freq         mean          std          min  \
car_url                   1          NaN          NaN          NaN   
complectation_dict       51          NaN          NaN          NaN   
equipment_dict          108          NaN          NaN          NaN   
image                    13          NaN          NaN          NaN   
modelDate               NaN      2007.07      7.41589         1904   
model_info             1404          NaN          NaN          NaN   
name                    631          NaN          NaN          NaN   
numberOfDoors           NaN      4.45082      0.70304            0   
parsing_unixtime        NaN  1.60329e+09       149307  1.60311e+09   
priceCurrency         34686          NaN          NaN          NaN   
sell_id                 NaN   1.0983e+09  1.91122e+07         2665   
super_gen               193          NaN          NaN          NaN   
vehicleConfiguration   2389          NaN          NaN          NaN   

                              25%          50%          75%          max  
car_url                       NaN          NaN          NaN          NaN  
complectation_dict            NaN          NaN          NaN          NaN  
equipment_dict                NaN          NaN          NaN          NaN  
image                         NaN          NaN          NaN          NaN  
modelDate                    2004         2008         2012         2020  
model_info                    NaN          NaN          NaN          NaN  
name                          NaN          NaN          NaN          NaN  
numberOfDoors                   4            5            5            5  
parsing_unixtime      1.60322e+09  1.60325e+09  1.60329e+09  1.60371e+09  
priceCurrency                 NaN          NaN          NaN          NaN  
sell_id               1.09905e+09  1.10091e+09  1.10125e+09  1.10137e+09  
super_gen                     NaN          NaN          NaN          NaN  
vehicleConfiguration          NaN          NaN          NaN          NaN

In [98]:
# check two features values
test[['enginePower', 'name']][:2]

enginePower                name
0     105 N12  1.2 AMT (105 л.с.)
1     110 N12   1.6 MT (110 л.с.)

In [99]:
# delete columns
train = train.drop(columns=['VIN', 'Гарантия', 'Госномер', 'Запас хода',
                            'Кузов №', 'Налог', 'Обмен', 'Статус', 'Unnamed: 0'])
test = test.drop(columns=['car_url', 'complectation_dict', 'equipment_dict', 'image', 'modelDate',
                          'model_info', 'name', 'numberOfDoors', 'parsing_unixtime',
                          'priceCurrency', 'sell_id', 'super_gen', 'vehicleConfiguration'])

## some add & replace
d) Bring brand name to **UPPER**

e) price delete **`₽`** & convert to numeric

f) create new columns `vednor` with two possible value **EUROPEAN & JAPANESE** -> lookup brand in test dataset in kaggle 

g) delete some row with "unusual" entries in brand names

In [100]:
# d)
train['brand'] = train.brand.apply(lambda x: x.upper())

In [101]:
# g) test brand names
train.groupby('brand')['brand'].count()

brand
*БЕЗ ДТП                                                                                                                                                                1
*ОФОРМЛЕНИЕ ЧЕРЕЗ ГАИ                                                                                                                                                   2
*ПРОЙДЕНЫ ВСЕ ТО                                                                                                                                                        1
*СЕРВИСНАЯ КНИЖКА                                                                                                                                                       1
AUDI                                                                                                                                                                 8183
BMW                                                                                                                                             

In [102]:
car_brand_list = ['skoda', 'audi', 'honda', 'volvo', 'bmw', 'nissan', 'infiniti',
                  'mercedes', 'toyota', 'lexus', 'volkswagen', 'mitsubishi']
car_brand_list = [x.upper() for x in car_brand_list]
train = train[train.brand.isin(car_brand_list)]
train.groupby('brand')['brand'].count()

brand
AUDI           8183
BMW            9925
HONDA          4831
INFINITI       1555
LEXUS          2483
MERCEDES       9512
MITSUBISHI     6245
NISSAN        12837
SKODA          7526
TOYOTA        13867
VOLKSWAGEN     9800
VOLVO          2904
Name: brand, dtype: int64

In [103]:
# e)
train['price'] = train['price'].str.replace('₽', '')
train = train[train['price'] != 'Рассмотрю варианты']
train['price'] = pd.to_numeric(train['price'])

In [104]:
# f)
EUROPEAN = ['AUDI', 'BMW', 'MERCEDES', 'SKODA', 'VOLKSWAGEN', 'VOLVO']
train['vendor'] = train.brand.apply(
    lambda x: 'EUROPEAN' if x in EUROPEAN else 'JAPANESE')

## model_name 
 h) `model_name` - car description in train dataset is more informative thatn in test and has another format datset as   
 - i'll bring model_name to the same format like kaggle

In [105]:
# h) bring model name into te same foramt / values in train dataset like in test
num_list = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
# test
#test['model_name'] = test.model_name.apply(lambda x: x.split('_')[0])
#test['model_name'] = test['model_name'].apply(lambda x: x[0] if x[0] in num_list else x)
# train
train['model_name'] = train['model_name'].apply(lambda x: x.upper())
train['model_name'] = [s.split(None, 1)[1] for s in train['model_name']]
train['model_name'] = train.model_name.apply(lambda x: x.split(' ')[0])
train['model_name'] = train.model_name.apply(lambda x: x.split('+')[0])
train['model_name'] = train.model_name.apply(
    lambda x: str(x).replace('-КЛАСС', '_KLASSE'))
train['model_name'] = train.model_name.apply(
    lambda x: str(x).replace('-', '_'))
train['model_name'] = train['model_name'].apply(
    lambda x: x[0] if x[0] in num_list else x)

In [106]:
for brand_name in train.brand.unique():
    print(brand_name, display(np.sort(test[test.brand == brand_name].model_name.unique()),
                              np.sort(train[train.brand == brand_name].model_name.unique())))

array(['02', '1ER', '2ACTIVETOURER', '2ER', '2GRANDTOURER', '321', '326',
       '340', '3ER', '4', '5ER', '6ER', '7ER', '8ER', 'E3', 'I3', 'I8',
       'M1', 'M2', 'M3', 'M4', 'M5', 'M6', 'X1', 'X2', 'X3', 'X3_M', 'X4',
       'X4_M', 'X5', 'X5_M', 'X6', 'X6_M', 'X7', 'Z1', 'Z3', 'Z3M', 'Z4'],
      dtype=object)

array(['0', '1', '2', '3', '4', '5', '6', '7', '8', 'I3', 'I8', 'M2',
       'M3', 'M4', 'M5', 'M6', 'M8', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6',
       'X7', 'Z3', 'Z4'], dtype=object)

BMW None


array(['100NX', '200SX', '280ZX', '300ZX', '350Z', 'AD', 'ALMERA',
       'ALMERA_CLASSIC', 'ALMERA_TINO', 'ALTIMA', 'ARMADA', 'AVENIR',
       'BASSARA', 'BLUEBIRD', 'BLUEBIRD_SYLPHY', 'CARAVAN_COACH',
       'CEDRIC', 'CEFIRO', 'CLIPPER', 'CUBE', 'DATSUN', 'DAYZ',
       'DAYZ_ROOX', 'ELGRAND', 'EXPERT', 'FAIRLADY_Z', 'FUGA', 'GLORIA',
       'GT_R', 'JUKE', 'JUKE_NISMO', 'KIX', 'LAFESTA', 'LARGO', 'LAUREL',
       'LEAF', 'LIBERTY', 'MARCH', 'MAXIMA', 'MICRA', 'MURANO', 'NAVARA',
       'NOTE', 'NP300', 'NV200', 'OTTI', 'PATHFINDER', 'PATROL',
       'PRAIRIE', 'PRESAGE', 'PRESEA', 'PRESIDENT', 'PRIMERA', 'PULSAR',
       'QASHQAI', 'QASHQAI_PLUS_2', 'QUEST', 'RNESSA', 'SAFARI', 'SENTRA',
       'SERENA', 'SILVIA', 'SKYLINE', 'STAGEA', 'SUNNY', 'TEANA',
       'TERRANO', 'TERRANO_REGULUS', 'TIIDA', 'TINO', 'URVAN', 'VANETTE',
       'WINGROAD', 'X_TRAIL'], dtype=object)

array(['1', '2', '3', 'AD', 'ALMERA', 'ALTIMA', 'ARMADA', 'AVENIR',
       'BASSARA', 'BLUEBIRD', 'CARAVAN', 'CEDRIC', 'CEFIRO', 'CHERRY',
       'CIMA', 'CLIPPER', 'CUBE', 'DATSUN', 'DAYZ', 'DUALIS', 'ELGRAND',
       'EXA', 'EXPERT', 'FAIRLADY', 'FUGA', 'GLORIA', 'GT_R', 'HOMY',
       'JUKE', 'KIX', 'LAFESTA', 'LANGLEY', 'LARGO', 'LATIO', 'LAUREL',
       'LEAF', 'LEOPARD', 'LIBERTY', 'LUCINO', 'MARCH', 'MAXIMA', 'MICRA',
       'MISTRAL', 'MOCO', 'MURANO', 'NAVARA', 'NOTE', 'NP300', 'NV100',
       'NV200', 'NV350', 'OTTI', 'PATHFINDER', 'PATROL', 'PINO', 'PIXO',
       'PRAIRIE', 'PRESAGE', 'PRESEA', 'PRESIDENT', 'PRIMASTAR',
       'PRIMERA', 'PULSAR', 'QASHQAI', 'QUEST', "R'NESSA", 'RASHEEN',
       'ROGUE', 'ROOX', 'SAFARI', 'SENTRA', 'SERENA', 'SILVIA', 'SKYLINE',
       'STAGEA', 'SUNNY', 'TEANA', 'TERRANO', 'TIIDA', 'TINO', 'TITAN',
       'URVAN', 'VANETTE', 'VERSA', 'WINGROAD', 'XTERRA', 'X_TRAIL'],
      dtype=object)

NISSAN None


array(['100_SERIES', '120', 'FABIA', 'FABIA_RS', 'FAVORIT', 'FELICIA',
       'KAROQ', 'KODIAQ', 'OCTAVIA', 'OCTAVIA_RS', 'POPULAR', 'RAPID',
       'ROOMSTER', 'SUPERB', 'YETI'], dtype=object)

array(['1', 'FABIA', 'FAVORIT', 'FELICIA', 'KAROQ', 'KODIAQ', 'OCTAVIA',
       'RAPID', 'ROOMSTER', 'SUPERB', 'YETI'], dtype=object)

SKODA None


array(['100', '200', '80', '90', '920', 'A1', 'A2', 'A3', 'A4',
       'A4_ALLROAD', 'A5', 'A6', 'A7', 'A8', 'ALLROAD', 'COUPE', 'E_TRON',
       'Q3', 'Q5', 'Q7', 'Q8', 'R8', 'RS3', 'RS4', 'RS5', 'RS6', 'RS7',
       'RSQ3', 'S3', 'S4', 'S5', 'S6', 'S7', 'S8', 'SQ5', 'TT', 'TTS',
       'TT_RS', 'V8'], dtype=object)

array(['1', '2', '8', '9', 'A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7', 'A8',
       'E_TRON', 'Q3', 'Q5', 'Q7', 'Q8', 'R8', 'RS', 'S3', 'S4', 'S5',
       'S6', 'S7', 'S8', 'SQ5', 'SQ7', 'TT', 'TTS', 'V8'], dtype=object)

AUDI None


array(['ACCORD', 'ACTY', 'AIRWAVE', 'ASCOT', 'AVANCIER', 'CITY', 'CIVIC',
       'CIVIC_FERIO', 'CIVIC_TYPE_R', 'CONCERTO', 'CROSSROAD',
       'CROSSTOUR', 'CR_V', 'CR_X', 'CR_Z', 'DOMANI', 'ELEMENT',
       'ELYSION', 'FIT', 'FREED', 'FR_V', 'HR_V', 'INSIGHT', 'INSPIRE',
       'INTEGRA', 'JAZZ', 'LEGEND', 'LOGO', 'MOBILIO', 'MOBILIO_SPIKE',
       'N_BOX', 'N_WGN', 'ODYSSEY', 'ODYSSEY_NA', 'ORTHIA', 'PARTNER',
       'PILOT', 'PRELUDE', 'RAFAGA', 'RIDGELINE', 'S2000', 'SABER',
       'SHUTTLE', 'STEPWAGON', 'STREAM', 'TORNEO', 'VEZEL', 'VIGOR', 'Z',
       'ZEST'], dtype=object)

array(['ACCORD', 'ACTY', 'AIRWAVE', 'ASCOT', 'AVANCIER', 'BEAT', 'CAPA',
       'CITY', 'CIVIC', 'CONCERTO', 'CROSSROAD', 'CROSSTOUR', 'CR_V',
       'CR_X', 'CR_Z', 'DOMANI', 'EDIX', 'ELEMENT', 'ELYSION', 'FIT',
       'FREED', 'FR_V', 'HR_V', 'INSIGHT', 'INSPIRE', 'INTEGRA', 'JADE',
       'JAZZ', 'LEGEND', 'LIFE', 'LOGO', 'MDX', 'MOBILIO', 'N_BOX',
       'N_WGN', 'ODYSSEY', 'ORTHIA', 'PARTNER', 'PASSPORT', 'PILOT',
       'PRELUDE', 'RAFAGA', 'RIDGELINE', 'S2000', 'SABER', 'SHUTTLE',
       'STEPWGN', 'STREAM', 'S_MX', "THAT'S", 'TORNEO', 'VAMOS', 'VEZEL',
       'VIGOR', 'Z', 'ZEST'], dtype=object)

HONDA None


array(['140', '240_SERIES', '440', '460', '740', '760', '850', '940',
       '960', 'C30', 'C70', 'S40', 'S60', 'S60_CROSS_COUNTRY', 'S70',
       'S80', 'S90', 'V40', 'V40_CC', 'V50', 'V60', 'V60_CROSS_COUNTRY',
       'V70', 'V90_CROSS_COUNTRY', 'XC40', 'XC60', 'XC70', 'XC90'],
      dtype=object)

array(['2', '3', '4', '7', '8', '9', 'C30', 'C70', 'S40', 'S60', 'S70',
       'S80', 'S90', 'V40', 'V50', 'V60', 'V70', 'V90', 'XC40', 'XC60',
       'XC70', 'XC90'], dtype=object)

VOLVO None


array(['EX', 'FX', 'G35', 'I', 'JX', 'M', 'Q', 'Q30', 'Q50', 'Q60', 'Q70',
       'QX30', 'QX4', 'QX50', 'QX56', 'QX60', 'QX70', 'QX80'],
      dtype=object)

array(['EX', 'FX', 'G', 'I', 'JX', 'M', 'Q', 'Q30', 'Q50', 'Q60', 'Q70',
       'QX30', 'QX4', 'QX50', 'QX56', 'QX60', 'QX70', 'QX80'],
      dtype=object)

INFINITI None


array(['190_SL', 'AMG_GLC_COUPE', 'AMG_GT', 'A_KLASSE', 'A_KLASSE_AMG',
       'B_KLASSE', 'CITAN', 'CLA_KLASSE', 'CLA_KLASSE_AMG', 'CLC_KLASSE',
       'CLK_KLASSE', 'CLS_KLASSE', 'CLS_KLASSE_AMG', 'CL_KLASSE',
       'CL_KLASSE_AMG', 'C_KLASSE', 'C_KLASSE_AMG', 'E_KLASSE',
       'E_KLASSE_AMG', 'GLA_CLASS', 'GLA_CLASS_AMG', 'GLB_KLASSE',
       'GLC_COUPE', 'GLC_KLASSE', 'GLC_KLASSE_AMG', 'GLE_KLASSE',
       'GLE_KLASSE_AMG', 'GLE_KLASSE_COUPE', 'GLE_KLASSE_COUPE_AMG',
       'GLK_KLASSE', 'GLS_KLASSE', 'GLS_KLASSE_AMG', 'GL_KLASSE',
       'GL_KLASSE_AMG', 'G_KLASSE', 'G_KLASSE_AMG', 'MAYBACH_GLS',
       'M_KLASSE', 'M_KLASSE_AMG', 'R_KLASSE', 'SIMPLEX', 'SLC_KLASSE',
       'SLK_KLASSE', 'SLR_KLASSE', 'SLS_AMG', 'SL_KLASSE',
       'SL_KLASSE_AMG', 'S_CLASS_MAYBACH', 'S_KLASSE', 'S_KLASSE_AMG',
       'VANEO', 'VIANO', 'VITO', 'V_KLASSE', 'W111', 'W114', 'W123',
       'W124', 'W128', 'W136', 'W186', 'W188', 'W201', 'X_KLASSE'],
      dtype=object)

array(['1', '2', 'AMG', 'A_KLASSE', 'B_KLASSE', 'CITAN', 'CLA',
       'CLC_KLASSE', 'CLK_KLASSE', 'CLS', 'CL_KLASSE', 'C_KLASSE', 'EQC',
       'E_KLASSE', 'GLA', 'GLB', 'GLC', 'GLE', 'GLK_KLASSE', 'GLS',
       'GL_KLASSE', 'G_KLASSE', 'MAYBACH', 'M_KLASSE', 'R_KLASSE',
       'SLK_KLASSE', 'SLS', 'SL_KLASSE', 'S_KLASSE', 'VANEO', 'VIANO',
       'VITO', 'V_KLASSE', 'W108', 'W111', 'W114', 'W115', 'W121', 'W123',
       'W124', 'W128', 'W136', 'W186', 'W188', 'W29', 'X_KLASSE'],
      dtype=object)

MERCEDES None


array(['4RUNNER', 'ALLEX', 'ALPHARD', 'ALTEZZA', 'AQUA', 'ARISTO',
       'AURIS', 'AVALON', 'AVENSIS', 'AVENSIS_VERSO', 'AYGO', 'BB',
       'BLADE', 'CALDINA', 'CAMI', 'CAMRY', 'CAMRY_SOLARA', 'CARINA',
       'CARINA_E', 'CARINA_ED', 'CAVALIER', 'CELICA', 'CHASER', 'COROLLA',
       'COROLLA_II', 'COROLLA_LEVIN', 'COROLLA_RUMION', 'COROLLA_SPACIO',
       'COROLLA_VERSO', 'CORONA', 'CORONA_EXIV', 'CORSA', 'CRESTA',
       'CROWN', 'CROWN_MAJESTA', 'CURREN', 'CYNOS', 'C_HR', 'DUET',
       'ECHO', 'ESQUIRE', 'ESTIMA', 'FJ_CRUISER', 'FORTUNER', 'FUNCARGO',
       'GAIA', 'GRAND_HIACE', 'GRANVIA', 'GT86', 'HARRIER', 'HIACE',
       'HIGHLANDER', 'HILUX', 'HILUX_SURF', 'IPSUM', 'IQ', 'ISIS', 'IST',
       'LAND_CRUISER', 'LAND_CRUISER_PRADO', 'LITE_ACE', 'MARK_II',
       'MARK_X', 'MARK_X_ZIO', 'MASTER_ACE_SURF', 'MATRIX',
       'MEGA_CRUISER', 'MR2', 'MR_S', 'NADIA', 'NOAH', 'OPA', 'ORIGIN',
       'PASEO', 'PASSO', 'PICNIC', 'PLATZ', 'PORTE', 'PREMIO', 'PREVIA',
       'PRIUS', 'PRI

array(['4', 'ALLEX', 'ALLION', 'ALPHARD', 'ALTEZZA', 'AQUA', 'ARISTO',
       'AURIS', 'AVALON', 'AVENSIS', 'AYGO', 'BB', 'BELTA', 'BLADE',
       'BREVIS', 'CALDINA', 'CAMI', 'CAMRY', 'CARINA', 'CAVALIER',
       'CELICA', 'CELSIOR', 'CENTURY', 'CHASER', 'COMS', 'COROLLA',
       'CORONA', 'CORSA', 'CRESTA', 'CROWN', 'CURREN', 'CYNOS', 'C_HR',
       'DUET', 'ECHO', 'ESQUIRE', 'ESTIMA', 'FJ', 'FORTUNER', 'FUNCARGO',
       'GAIA', 'GRAND', 'GRANVIA', 'GT86', 'HARRIER', 'HIACE',
       'HIGHLANDER', 'HILUX', 'IPSUM', 'IQ', 'ISIS', 'IST', 'KLUGER',
       'LAND', 'LITE', 'MARK', 'MASTERACE', 'MATRIX', 'MEGA', 'MR2',
       'MR_S', 'NADIA', 'NOAH', 'OPA', 'PASSO', 'PICNIC', 'PIXIS',
       'PLATZ', 'PORTE', 'PREMIO', 'PREVIA', 'PRIUS', 'PROBOX', 'PROGRES',
       'PRONARD', 'RACTIS', 'RAUM', 'RAV4', 'REGIUS', 'ROOMY', 'RUSH',
       'SAI', 'SCEPTER', 'SEQUOIA', 'SIENNA', 'SIENTA', 'SOARER', 'SPADE',
       'SPARKY', 'SPRINTER', 'STARLET', 'SUCCEED', 'SUPRA', 'TACOMA',
       'TANK', 'TER

TOYOTA None


array(['AMAROK', 'ARTEON', 'ATLAS', 'BEETLE', 'BORA', 'CADDY',
       'CALIFORNIA', 'CARAVELLE', 'CORRADO', 'EOS', 'EUROVAN', 'FOX',
       'GOLF', 'GOLF_GTI', 'GOLF_PLUS', 'GOLF_R', 'GOLF_R32', 'JETTA',
       'LUPO', 'MULTIVAN', 'PASSAT', 'PASSAT_CC', 'PHAETON', 'POINTER',
       'POLO', 'POLO_GTI', 'SCIROCCO', 'SHARAN', 'TARO', 'TERAMONT',
       'TIGUAN', 'TOUAREG', 'TOURAN', 'TRANSPORTER', 'TYPE_1', 'VENTO'],
      dtype=object)

array(['AMAROK', 'ARTEON', 'BEETLE', 'BORA', 'CADDY', 'CALIFORNIA',
       'CARAVELLE', 'CORRADO', 'FOX', 'GOLF', 'ILTIS', 'JETTA', 'LUPO',
       'MULTIVAN', 'PASSAT', 'PHAETON', 'POINTER', 'POLO', 'SANTANA',
       'SCIROCCO', 'SHARAN', 'TERAMONT', 'TIGUAN', 'TOUAREG', 'TOURAN',
       'TRANSPORTER', 'TYPE', 'VENTO'], dtype=object)

VOLKSWAGEN None


array(['CT', 'ES', 'GS', 'GX', 'HS', 'IS', 'LM', 'LS', 'LX', 'NX', 'RC',
       'RX', 'SC', 'UX'], dtype=object)

array(['CT', 'ES', 'GS', 'GX', 'IS', 'LC', 'LM', 'LS', 'LX', 'NX', 'RC',
       'RX', 'SC', 'UX'], dtype=object)

LEXUS None


array(['AIRTREK', 'ASPIRE', 'ASX', 'CARISMA', 'CHARIOT', 'COLT', 'DELICA',
       'DELICA_D2', 'DELICA_D_5', 'DIAMANTE', 'DINGO', 'DION', 'ECLIPSE',
       'ECLIPSE_CROSS', 'EK_WAGON', 'EMERAUDE', 'ENDEAVOR', 'FTO',
       'GALANT', 'GRANDIS', 'I_MIEV', 'L200', 'L300', 'L400', 'LANCER',
       'LANCER_EVOLUTION', 'LANCER_RALLIART', 'LEGNUM', 'LIBERO',
       'MINICA', 'MINICAB', 'MIRAGE', 'MONTERO', 'MONTERO_SPORT',
       'OUTLANDER', 'PAJERO', 'PAJERO_IO', 'PAJERO_MINI', 'PAJERO_PININ',
       'PAJERO_SPORT', 'RVR', 'SIGMA', 'SPACE_GEAR', 'SPACE_RUNNER',
       'SPACE_STAR', 'SPACE_WAGON'], dtype=object)

array(['3', 'AIRTREK', 'ASPIRE', 'ASX', 'CARISMA', 'CHALLENGER',
       'CHARIOT', 'COLT', 'DEBONAIR', 'DELICA', 'DIAMANTE', 'DINGO',
       'DION', 'ECLIPSE', 'EK', 'EMERAUDE', 'ENDEAVOR', 'ETERNA', 'FTO',
       'GALANT', 'GRANDIS', 'GTO', 'I', 'I_MIEV', 'L200', 'L300', 'L400',
       'LANCER', 'LEGNUM', 'LIBERO', 'MINICA', 'MINICAB', 'MIRAGE',
       'MONTERO', 'OUTLANDER', 'PAJERO', 'RVR', 'SPACE', 'TOPPO'],
      dtype=object)

MITSUBISHI None


In [107]:
# check values in featurs in test & train datasets

In [108]:
#train['productionDate'] = train.productionDate.apply(lambda x: str(x).replace('.', ''))
train['productionDate'] = pd.to_numeric(train.productionDate)

## functions to print and compare datasets values

to reduce the code I created a function to comare two dataset values 

In [110]:
# function tocheck values in features
def feature_values_compare(feature):
    print('test \n', test[feature].unique(),
          '\n train \n',  train[feature].unique(),
          '\n test: \n',  test[feature].dtypes,
          '\n train: \n', train[feature].dtypes,)

    print('train', set(train[feature].unique()).difference(
        set(test[feature].unique())))
    print('test', set(test[feature].unique()).difference(
        set(train[feature].unique())))

# function tocheck values in features


def feature_values_print(feature):
    print('test \n', test[feature].unique(),
          '\n train \n',  train[feature].unique(),
          '\n test: \n',  test[feature].dtypes,
          '\n train: \n', train[feature].dtypes,)

### brand 
brand has the same values like in test
- nothing TODO

In [111]:
# brand
feature_values_compare('brand')

test 
 ['SKODA' 'AUDI' 'HONDA' 'VOLVO' 'BMW' 'NISSAN' 'INFINITI' 'MERCEDES'
 'TOYOTA' 'LEXUS' 'VOLKSWAGEN' 'MITSUBISHI'] 
 train 
 ['BMW' 'NISSAN' 'SKODA' 'AUDI' 'HONDA' 'VOLVO' 'INFINITI' 'MERCEDES'
 'TOYOTA' 'VOLKSWAGEN' 'LEXUS' 'MITSUBISHI'] 
 test: 
 object 
 train: 
 object
train set()
test set()


### productionDate
- bring to the same dtype (train - float, test - int64)

In [112]:
# productionDate
feature_values_compare('productionDate')

test 
 [2014 2017 2012 2011 2019 2018 2010 2020 2016 2013 2006 2007 2015 2005
 2008 2009 1997 2004 2002 1987 2003 2001 1976 2000 1998 1995 1999 1993
 1939 1996 1984 1990 1991 1992 1989 1982 1985 1994 1938 1981 1988 1983
 1980 1986 1978 1970 1979 1977 1972 1975 1969 1950 1953 1949 1937 1959
 1968 1936 1904 1974 1967 1961 1960 1965 1963 1957 1952 1973 1948] 
 train 
 [1988. 2013. 2020. 2018. 2019. 2010. 2012. 2017. 2015. 2002. 2009. 2016.
 2008. 2014. 2001. 1985. 2011. 1999. 2005. 2003. 2007. 2006. 1998. 1991.
 1997. 1992. 1990. 2004. 2000. 1993. 1994. 1987. 1996. 1995. 1984. 1986.
 1989. 1979. 2021. 1982. 1983. 1978. 1981. 1980. 1974. 1973. 1977. 1937.
 1976. 1960. 1965. 1940. 1934. 1938. 1964. 1972. 1959. 1968. 1950. 1948.
 1952. 1967. 1969. 1975. 1961. 1953. 1957. 1963. 1949. 1939. 1958.] 
 test: 
 int64 
 train: 
 float64
train {2021.0, 1958.0, 1964.0, 1934.0, 1940.0}
test {1936, 1970, 1904}


In [113]:
# bring float to int
train['productionDate'] = train.productionDate.astype(int)

In [114]:
# check
feature_values_compare('productionDate')

test 
 [2014 2017 2012 2011 2019 2018 2010 2020 2016 2013 2006 2007 2015 2005
 2008 2009 1997 2004 2002 1987 2003 2001 1976 2000 1998 1995 1999 1993
 1939 1996 1984 1990 1991 1992 1989 1982 1985 1994 1938 1981 1988 1983
 1980 1986 1978 1970 1979 1977 1972 1975 1969 1950 1953 1949 1937 1959
 1968 1936 1904 1974 1967 1961 1960 1965 1963 1957 1952 1973 1948] 
 train 
 [1988 2013 2020 2018 2019 2010 2012 2017 2015 2002 2009 2016 2008 2014
 2001 1985 2011 1999 2005 2003 2007 2006 1998 1991 1997 1992 1990 2004
 2000 1993 1994 1987 1996 1995 1984 1986 1989 1979 2021 1982 1983 1978
 1981 1980 1974 1973 1977 1937 1976 1960 1965 1940 1934 1938 1964 1972
 1959 1968 1950 1948 1952 1967 1969 1975 1961 1953 1957 1963 1949 1939
 1958] 
 test: 
 int64 
 train: 
 int64
train {2021, 1958, 1964, 1934, 1940}
test {1936, 1970, 1904}


### mileage
-  replace **`\xa0`** and км bring to **int** dtype

In [115]:
#mileage
feature_values_print('mileage')

test 
 [ 74000  60563  88000 ... 121276 212678 157965] 
 train 
 ['226\xa0000\xa0км' '75\xa0000\xa0км' '18\xa0565\xa0км' ...
 '130\xa0478\xa0км' '93\xa0199\xa0км' '169\xa0805\xa0км'] 
 test: 
 int64 
 train: 
 object


In [116]:
# replace some sighns and bring to the same type
train['mileage'] = train.mileage.apply(lambda x: str(x).replace('\xa0', ''))
train['mileage'] = train.mileage.apply(lambda x: str(x).replace('км', ''))
train['mileage'] = train.mileage.astype(int)
# check
feature_values_print('mileage')

test 
 [ 74000  60563  88000 ... 121276 212678 157965] 
 train 
 [226000  75000  18565 ... 130478  93199 169805] 
 test: 
 int64 
 train: 
 int64


### bodyType
- except 3 values all bodyTape values are equal. 
- That concerns only 5 vlaues (1 in test and 4 in train dataset). 
- I will not delete this values besause in other feature could be important inforamtion and i do not have to many data to delete rows

In [117]:
# bodyType
feature_values_compare('bodyType')

test 
 ['лифтбек' 'внедорожник 5 дв.' 'хэтчбек 5 дв.' 'седан' 'компактвэн'
 'универсал 5 дв.' 'пикап одинарная кабина' 'хэтчбек 3 дв.' 'купе'
 'кабриолет' 'минивэн' 'пикап двойная кабина' 'внедорожник 3 дв.'
 'родстер' 'микровэн' 'седан 2 дв.' 'купе-хардтоп' 'фастбек' 'тарга'
 'внедорожник открытый' 'лимузин' 'пикап полуторная кабина'
 'седан-хардтоп' 'фургон'] 
 train 
 ['седан' 'внедорожник 5 дв.' 'лифтбек' 'купе' 'хэтчбек 5 дв.'
 'хэтчбек 3 дв.' 'компактвэн' 'пикап двойная кабина' 'универсал 5 дв.'
 'минивэн' 'пикап полуторная кабина' 'внедорожник 3 дв.' 'микровэн'
 'кабриолет' 'родстер' 'седан 2 дв.' 'тарга' 'фургон' 'купе-хардтоп'
 'внедорожник открытый' 'седан-хардтоп' 'пикап одинарная кабина'
 'универсал 3 дв.' nan 'лимузин'] 
 test: 
 object 
 train: 
 object
train {nan, 'универсал 3 дв.'}
test {'фастбек'}


In [118]:
# check this values in test dataset
test[test['bodyType'] == 'фастбек']

bodyType   brand        color  \
11764  фастбек  NISSAN  серебристый   

                                             description engineDisplacement  \
11764  Легендарное спорт купе Ниссан 280zx, выпущен в...            2.8 LTR   

      enginePower fuelType  mileage model_name  productionDate  \
11764     145 N12   бензин     2000      280ZX            1980   

      vehicleTransmission    vendor    Владельцы Владение       ПТС  Привод  \
11764        механическая  JAPANESE  2 владельца      NaN  Дубликат  задний   

         Руль           Состояние     Таможня  
11764  Правый  Не требует ремонта  Растаможен

In [119]:
#check in train dataset
train[(train['bodyType'] == 'nan') |(train['bodyType'] == 'универсал 3 дв.')]

brand model_name   price  productionDate  mileage         bodyType  \
835   SKODA          1  400000            1960    18000  универсал 3 дв.   
831  NISSAN         AD  120000            1996    30000  универсал 3 дв.   
293   SKODA          1  150000            1958    76000  универсал 3 дв.   
151   SKODA          1  400000            1960    18000  универсал 3 дв.   

       color engineDisplacement vehicleTransmission    Привод    Руль  \
835  голубой            1.2 LTR        механическая    задний   Левый   
831    белый            1.5 LTR        механическая  передний  Правый   
293  голубой            1.2 LTR        механическая    задний   Левый   
151  голубой            1.2 LTR        механическая    задний   Левый   

              Состояние    Владельцы       ПТС     Таможня  \
835  Не требует ремонта   1 владелец  Оригинал  Растаможен   
831  Не требует ремонта  3 или более  Оригинал  Растаможен   
293  Не требует ремонта   1 владелец  Оригинал  Растаможен   
151  Не требует ремонта   1 владелец  Оригинал  Растаможен   

                                           description            Владение  \
835  Шкода «Стейшн» 1201 «санитарка»1960 год выпуск...                 NaN   
831                           Дизель после кап ремонта   2 года и 4 месяца   
293  Шкода 1958г выпуска,мотор кпп оригинал.заведёт...  62 года и 4 месяца   
151  Шкода «Стейшн» 1201 «санитарка»1960 год выпуск...                 NaN   

    enginePower fuelType    vendor  
835     45  N12   Бензин  EUROPEAN  
831     94  N12   Бензин  JAPANESE  
293     45  N12   Бензин  EUROPEAN  
151     45  N12   Бензин  EUROPEAN

### color
- `color` values in both datasets are the same

In [120]:
#color
feature_values_compare('color')

test 
 ['синий' 'чёрный' 'серый' 'коричневый' 'белый' 'пурпурный' 'бежевый'
 'серебристый' 'красный' 'зелёный' 'жёлтый' 'голубой' 'оранжевый'
 'фиолетовый' 'золотистый' 'розовый'] 
 train 
 ['чёрный' 'белый' 'серый' 'серебристый' 'синий' 'коричневый' 'бежевый'
 'красный' 'фиолетовый' 'зелёный' 'золотистый' 'оранжевый' 'пурпурный'
 'голубой' 'жёлтый' 'розовый'] 
 test: 
 object 
 train: 
 object
train set()
test set()


In [ ]:
#engineDisplacement
feature_values_compare('engineDisplacement')

### engineDisplacement
-`engineDisplacement` in train dataset the values differ from test - bring to the same format  

In [ ]:
feature_values_compare('engineDisplacement')

In [121]:
# replace some characters to bring to the same from like in test dataset
train['engineDisplacement'] = train.engineDisplacement.apply(
    lambda x: str(x).replace('\xa0', ''))
train['engineDisplacement'] = train.engineDisplacement.apply(
    lambda x: str(x).replace('.с.', ''))
train['engineDisplacement'] = train['engineDisplacement'].str[:1] + \
    '.' + train['engineDisplacement'].str[1:]
train['engineDisplacement'] = train.engineDisplacement.apply(
    lambda x: str(x).replace('..', '.'))
train['engineDisplacement'] = train['engineDisplacement'].str[:3]
train['engineDisplacement'] = train['engineDisplacement'].str[:] + ' LTR'


feature_values_compare('engineDisplacement')

test 
 ['1.2 LTR' '1.6 LTR' '1.8 LTR' '2.0 LTR' '1.4 LTR' '1.3 LTR' '1.0 LTR'
 '3.6 LTR' '1.5 LTR' '1.9 LTR' '2.8 LTR' '1.1 LTR' '2.5 LTR' '4.2 LTR'
 '3.0 LTR' '4.0 LTR' '5.9 LTR' '2.7 LTR' '3.1 LTR' '2.4 LTR' '5.2 LTR'
 '3.2 LTR' '4.1 LTR' '6.3 LTR' '2.3 LTR' '6.0 LTR' '2.2 LTR' '3.7 LTR'
 '2.9 LTR' '5.0 LTR' '3.3 LTR' '2.1 LTR' '2.6 LTR' ' LTR' '3.5 LTR'
 '1.7 LTR' '0.7 LTR' '4.4 LTR' '4.8 LTR' '5.4 LTR' '6.6 LTR' '4.9 LTR'
 '3.8 LTR' '3.4 LTR' '3.9 LTR' '4.6 LTR' '5.6 LTR' '4.5 LTR' '5.5 LTR'
 '6.2 LTR' '4.7 LTR' '4.3 LTR' '5.8 LTR' '5.3 LTR' '5.7 LTR'] 
 train 
 ['2.5 LTR' '4.4 LTR' '2.0 LTR' '3.0 LTR' '5.0 LTR' '1.8 LTR' '1.6 LTR'
 '4.6 LTR' '3.4 LTR' '2.2 LTR' '3.5 LTR' '1.5 LTR' '1.4 LTR' '5.6 LTR'
 '1.0 LTR' '1.3 LTR' '1.2 LTR' '2.8 LTR' '2.4 LTR' '4.5 LTR' '4.2 LTR'
 '2.7 LTR' '3.3 LTR' '4.0 LTR' '4.1 LTR' '0.7 LTR' '1.9 LTR' '5.2 LTR'
 '5.9 LTR' '2.3 LTR' '3.2 LTR' '3.1 LTR' '6.0 LTR' '3.6 LTR' '3.7 LTR'
 '2.6 LTR' '1.7 LTR' '2.9 LTR' '4.8 LTR' '5.4 LTR' '2.1 LTR' '4.7 LTR'
 

### vehicleTransmission, Привод, Руль, Владельцы & ПТС
- `vehicleTransmission`, `Привод`, `, Руль` , `Владельцы`, `ПТС` nothing to add, some nan's but it will be done in EDA

In [122]:
# vehicleTransmission
feature_values_compare('vehicleTransmission')

test 
 ['роботизированная' 'механическая' 'автоматическая' 'вариатор'] 
 train 
 ['механическая' 'автоматическая' 'роботизированная' 'вариатор' nan] 
 test: 
 object 
 train: 
 object
train {nan}
test set()


In [123]:
# Привод
feature_values_compare('Привод')

test 
 ['передний' 'полный' 'задний'] 
 train 
 ['задний' 'полный' 'передний' nan] 
 test: 
 object 
 train: 
 object
train {nan}
test set()


In [124]:
# Руль
feature_values_compare('Руль')

test 
 ['Левый' 'Правый'] 
 train 
 ['Левый' 'Правый'] 
 test: 
 object 
 train: 
 object
train set()
test set()


In [125]:
#Владельцы
feature_values_compare('Владельцы')

test 
 ['3 или более' '1\xa0владелец' '2\xa0владельца'] 
 train 
 ['3 или более' '2\xa0владельца' '1\xa0владелец' nan] 
 test: 
 object 
 train: 
 object
train {nan}
test set()


In [126]:
# ПТС
feature_values_compare('ПТС')

test 
 ['Оригинал' 'Дубликат' nan] 
 train 
 ['Оригинал' 'Дубликат' nan] 
 test: 
 object 
 train: 
 object
train set()
test set()


### Состояние & Таможня
- `Состояние`, `Таможня` in train we have two values more, and in test only one, it could have big influence on the model, maybe it make sense to deelte the feature

In [127]:
# Состояние
feature_values_compare('Состояние')

test 
 ['Не требует ремонта'] 
 train 
 ['Не требует ремонта' 'Битый / не на ходу'] 
 test: 
 object 
 train: 
 object
train {'Битый / не на ходу'}
test set()


In [128]:
# Таможня
feature_values_compare('Таможня')

test 
 ['Растаможен'] 
 train 
 ['Растаможен' 'Не растаможен'] 
 test: 
 object 
 train: 
 object
train {'Не растаможен'}
test set()


### Владение
- `Владение` most of values have the same vealues and same format

In [129]:
# Владение
feature_values_compare('Владение')

test 
 [nan '3 года и 2 месяца' '11 месяцев' '4 года и 1 месяц'
 '2 года и 4 месяца' '1 месяц' '3 года и 10 месяцев' '2 года и 6 месяцев'
 '6 лет и 4 месяца' '2 месяца' '4 месяца' '1 год и 10 месяцев'
 '2 года и 7 месяцев' '7 лет и 11 месяцев' '7 лет и 5 месяцев'
 '1 год и 11 месяцев' '2 года и 1 месяц' '7 лет' '4 года и 2 месяца'
 '2 года и 8 месяцев' '2 года и 5 месяцев' '12 лет и 11 месяцев'
 '5 лет и 11 месяцев' '2 года и 11 месяцев' '3 года и 1 месяц'
 '1 год и 1 месяц' '1 год' '1 год и 7 месяцев' '3 месяца'
 '1 год и 5 месяцев' '1 год и 2 месяца' '7 лет и 8 месяцев'
 '3 года и 4 месяца' '1 год и 6 месяцев' '12 лет и 8 месяцев'
 '6 лет и 1 месяц' '6 лет и 11 месяцев' '5 лет и 4 месяца'
 '3 года и 5 месяцев' '6 лет и 2 месяца' '4 года и 6 месяцев'
 '3 года и 6 месяцев' '1 год и 8 месяцев' '1 год и 9 месяцев'
 '3 года и 11 месяцев' '2 года и 2 месяца' '3 года' '5 лет и 5 месяцев'
 '2 года и 9 месяцев' '6 лет и 9 месяцев' '9 месяцев' '1 год и 3 месяца'
 '4 года и 3 месяца' '8 месяцев

### enginePower
- `enginePower` bring to the same format

In [130]:
# enginePower
feature_values_print('enginePower')

test 
 ['105 N12' '110 N12' '152 N12' '200 N12' '102 N12' '150 N12' '90 N12'
 '180 N12' '220 N12' '122 N12' '70 N12' '140 N12' '125 N12' '54 N12'
 '86 N12' '75 N12' '64 N12' '95 N12' '260 N12' '170 N12' '80 N12' '68 N12'
 '160 N12' '115 N12' '280 N12' '53 N12' '60 N12' '143 N12' '42 N12'
 '101 N12' '58 N12' '193 N12' '79 N12' '30 N12' '100 N12' '50 N12'
 '163 N12' '225 N12' '420 N12' '211 N12' '245 N12' '560 N12' '500 N12'
 '249 N12' '450 N12' '605 N12' '250 N12' '354 N12' '120 N12' '290 N12'
 '230 N12' '350 N12' '204 N12' '255 N12' '340 N12' '177 N12' '272 N12'
 '372 N12' '210 N12' '130 N12' '300 N12' '190 N12' '239 N12' '435 N12'
 '333 N12' '271 N12' '326 N12' '238 N12' '310 N12' '233 N12' '252 N12'
 '133 N12' '460 N12' '520 N12' '400 N12' '525 N12' '367 N12' '265 N12'
 '550 N12' '580 N12' '88 N12' '165 N12' '430 N12' '335 N12' '174 N12'
 '224 N12' '136 N12' '256 N12' '338 N12' '112 N12' '218 N12' '360 N12'
 '93 N12' '182 N12' '113 N12' '240 N12' '270 N12' '137 N12' '155 N12'
 '61 N1

In [131]:
train['enginePower'] = train.enginePower.apply(
    lambda x: str(x).replace('\xa0', ''))
train['enginePower'] = train.enginePower.apply(
    lambda x: str(x).replace('кВт', ' N12'))

feature_values_compare('enginePower')

test 
 ['105 N12' '110 N12' '152 N12' '200 N12' '102 N12' '150 N12' '90 N12'
 '180 N12' '220 N12' '122 N12' '70 N12' '140 N12' '125 N12' '54 N12'
 '86 N12' '75 N12' '64 N12' '95 N12' '260 N12' '170 N12' '80 N12' '68 N12'
 '160 N12' '115 N12' '280 N12' '53 N12' '60 N12' '143 N12' '42 N12'
 '101 N12' '58 N12' '193 N12' '79 N12' '30 N12' '100 N12' '50 N12'
 '163 N12' '225 N12' '420 N12' '211 N12' '245 N12' '560 N12' '500 N12'
 '249 N12' '450 N12' '605 N12' '250 N12' '354 N12' '120 N12' '290 N12'
 '230 N12' '350 N12' '204 N12' '255 N12' '340 N12' '177 N12' '272 N12'
 '372 N12' '210 N12' '130 N12' '300 N12' '190 N12' '239 N12' '435 N12'
 '333 N12' '271 N12' '326 N12' '238 N12' '310 N12' '233 N12' '252 N12'
 '133 N12' '460 N12' '520 N12' '400 N12' '525 N12' '367 N12' '265 N12'
 '550 N12' '580 N12' '88 N12' '165 N12' '430 N12' '335 N12' '174 N12'
 '224 N12' '136 N12' '256 N12' '338 N12' '112 N12' '218 N12' '360 N12'
 '93 N12' '182 N12' '113 N12' '240 N12' '270 N12' '137 N12' '155 N12'
 '61 N1

### fuelType
- `fuelType` the values in train and tst dataset differ. There are many cars with two fuel types. I deleted the information about **газобаллонное оборудование** to bring the values to the same format

In [132]:
# fuelType
feature_values_compare('fuelType')

test 
 ['бензин' 'дизель' 'гибрид' 'электро' 'газ'] 
 train 
 [' Бензин' ' Дизель' ' Бензин, газобаллонное оборудование' ' Электро'
 ' Гибрид' ' Газ' ' Гибрид, газобаллонное оборудование'
 ' Дизель, газобаллонное оборудование' ' Газ, газобаллонное оборудование'
 nan] 
 test: 
 object 
 train: 
 object
train {nan, ' Бензин', ' Электро', ' Газ, газобаллонное оборудование', ' Гибрид', ' Гибрид, газобаллонное оборудование', ' Газ', ' Дизель', ' Бензин, газобаллонное оборудование', ' Дизель, газобаллонное оборудование'}
test {'бензин', 'дизель', 'электро', 'газ', 'гибрид'}


In [133]:
train['fuelType'] = train['fuelType'].str.lower()
train['fuelType'] = train.fuelType.apply(
    lambda x: str(x).replace(', газобаллонное оборудование', ''))
train['fuelType'] = train.fuelType.apply(lambda x: str(x).replace(' ', ''))
feature_values_compare('fuelType')

test 
 ['бензин' 'дизель' 'гибрид' 'электро' 'газ'] 
 train 
 ['бензин' 'дизель' 'электро' 'гибрид' 'газ' 'nan'] 
 test: 
 object 
 train: 
 object
train {'nan'}
test set()


### vendor

In [ ]:
# vendor
feature_values_compare('vendor')

In [135]:
train.to_csv('data_for_kaggle.csv', index=False)

In [134]:
train

brand model_name    price  productionDate  mileage           bodyType  \
0     BMW          5   155000            1988   226000              седан   
1     BMW         X6  1950000            2013    75000  внедорожник 5 дв.   
2     BMW         X1  2800000            2020    18565  внедорожник 5 дв.   
3     BMW         X4  2950000            2018    21000  внедорожник 5 дв.   
4     BMW         X5  5450000            2019    78200  внедорожник 5 дв.   
..    ...        ...      ...             ...      ...                ...   
842  AUDI         Q3  1364000            2013    93199  внедорожник 5 дв.   
843  AUDI         Q7  3120000            2015    96717  внедорожник 5 дв.   
844  AUDI         A6  1629000            2016   107000              седан   
845  AUDI         Q5  1049000            2011   169805  внедорожник 5 дв.   
846  AUDI         A4   945000            2013   170000              седан   

          color engineDisplacement vehicleTransmission    Привод   Руль  \
0        чёрный            2.5 LTR        механическая    задний  Левый   
1         белый            4.4 LTR      автоматическая    полный  Левый   
2         белый            2.0 LTR      автоматическая    полный  Левый   
3         серый            2.0 LTR      автоматическая    полный  Левый   
4        чёрный            3.0 LTR      автоматическая    полный  Левый   
..          ...                ...                 ...       ...    ...   
842       серый            2.0 LTR    роботизированная    полный  Левый   
843  коричневый            3.0 LTR      автоматическая    полный  Левый   
844      чёрный            1.8 LTR    роботизированная  передний  Левый   
845       белый            2.0 LTR    роботизированная    полный  Левый   
846      чёрный            2.0 LTR    роботизированная    полный  Левый   

              Состояние    Владельцы       ПТС     Таможня  \
0    Не требует ремонта  3 или более  Оригинал  Растаможен   
1    Не требует ремонта  2 владельца  Оригинал  Растаможен   
2    Не требует ремонта   1 владелец  Оригинал  Растаможен   
3    Не требует ремонта  3 или более  Дубликат  Растаможен   
4    Не требует ремонта   1 владелец  Оригинал  Растаможен   
..                  ...          ...       ...         ...   
842  Не требует ремонта  2 владельца  Оригинал  Растаможен   
843  Не требует ремонта   1 владелец  Оригинал  Растаможен   
844  Не требует ремонта  2 владельца  Оригинал  Растаможен   
845  Не требует ремонта  3 или более  Оригинал  Растаможен   
846  Не требует ремонта  3 или более  Дубликат  Растаможен   

                                           description           Владение  \
0    Продаю авто, за время владения было не мало чт...                NaN   
1    Дилерский авто , в матовой  защитной пленке  (...  6 лет и 9 месяцев   
2    Активная безопасность:— Антиблокировочная сист...                NaN   
3    БМВ-Х4. Бензин -245л.сил. АКПП-8. Куплен у ОФ....                NaN   
4    Куплен и oбcлуживaлся у официального дилерa. О...   1 год и 2 месяца   
..                                                 ...                ...   
842  Ауди Q3 в отличном состоянии.Пробег родной, по...    1 год и 1 месяц   
843  Гаражный номер: КД-21025-Р\r КД-21025-Р. 1 соб...                NaN   
844  Ваш выбор и Ваше время — наши главные приорите...                NaN   
845  Audi Q5 Quattro2011 г.в.,Пробег - 169 805 км.,...                NaN   
846  Доброго времени суток всем! Продаю свой любимы...                NaN   

    enginePower fuelType    vendor  
0       170 N12   бензин  EUROPEAN  
1       555 N12   бензин  EUROPEAN  
2       150 N12   дизель  EUROPEAN  
3       245 N12   бензин  EUROPEAN  
4       249 N12   дизель  EUROPEAN  
..          ...      ...       ...  
842     170 N12   бензин  EUROPEAN  
843     249 N12   дизель  EUROPEAN  
844     190 N12   бензин  EUROPEAN  
845     211 N12   бензин  EUROPEAN  
846     225 N12   бензин  EUROPEAN  

[89668 rows x 20 columns]